subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyHandler.java

```diff
@@ -127,10 +128,17 @@ private Dependency doAdd(Configuration configuration, Object dependencyNotation,
             configuration.extendsFrom(other);
             return null;
         }
-
-        Dependency dependency = create(dependencyNotation, configureClosure);
-        configuration.getDependencies().add(dependency);
-        return dependency;
+        if (dependencyNotation instanceof Provider<?>) {
+            Provider<?> lazy = (Provider<?>) dependencyNotation;
+            Provider<Dependency> lazyDependency = lazy.map(lazyNotation -> create(lazyNotation, configureClosure));
+            configuration.getDependencies().addLater(lazyDependency);
```

Comment:
We've been successfully able to use `addLater` in the Ktlint Gradle plugin here.

https://github.com/JLLeitschuh/ktlint-gradle/blob/fe91e0ec6bd841dd0c351d97f971cc6d7b450fc4/plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/Configurations.kt#L32